### PR TITLE
Add dam com1 dfa

### PR DESCRIPTION
**Pull request by matthiasto**
Split the cython files (this required to create some .pxd files)
Add the dam tool

To test it, avaSlide, avaBowl and avaKot have some dam Input data.
Activate the dam in the com1DFA ini file and run.

To compare to com1DFAOrig, choose the correct line (the one corresponding to the parabola, bowl or kot) in the createProject.cint
Run com1DFAOrig, then run com1DFA (using the same particles) and finally run the ana3AIMECComp

The documentation is on it's way.

### CODE REVIEW

**Comment by fso42** [comment]
'test_damCom1DFA.py' fails as the *WallLines functions expect wallLineDict['fileName'] . But as far as I can see this is only for log.infos?

File: avaframe/com1DFA/DFAfunctionsCython.pyx [hunk 1]
```diff
@@ -720,49 +667,104 @@ def updatePositionC(cfg, particles, dem, force, int typeStop=0):
...
+        # if distReproj:
+        #   xNew, yNew, zNew, iCellNew, LxNew0, LyNew0, wNew[0], wNew[1], wNew[2], wNew[3] = distConservProjectionIteratrive(
+        #     x, y, z, ZDEM, nxArray, nyArray, nzArray, xNew, yNew, zNew, csz, ncols, nrows, interpOption,
+        #     reprojectionIterations, thresholdProjection)
+        # else:
+        #   xNew, yNew, iCellNew, LxNew0, LyNew0, wNew[0], wNew[1], wNew[2], wNew[3] = samosProjectionIteratrive(
+        #     xNew, yNew, zNew, ZDEM, nxArray, nyArray, nzArray, csz, ncols, nrows, interpOption, reprojectionIterations)
+        #   if iCellNew >= 0:
+        #     zNew = getScalar(LxNew0, LyNew0, wNew[0], wNew[1], wNew[2], wNew[3], ZDEM)
+
```
Comment (fso42) [inline avaframe/com1DFA/DFAfunctionsCython.pyx:721]: remove this?

File: avaframe/com1DFA/DFAfunctionsCython.pyx [hunk 2]
```diff
@@ -720,49 +667,104 @@ def updatePositionC(cfg, particles, dem, force, int typeStop=0):
...
-      uyNew = uyNew * uMag / (uMagNew + velMagMin)
-      uzNew = uzNew * uMag / (uMagNew + velMagMin)
+
+    # reproject velocity
+    uxNew, uyNew, uzNew, uMag = DFAtlsC.reprojectVelocity(uxNew, uyNew, uzNew, nxNew, nyNew, nzNew, velMagMin)
+
+    #  ############## Start add Dam interaction ##############################################
+    if flagDam and cellsCrossed[iCell] == 1:
+      # if there is an interaction with the dam, update position and velocity
+      inter, xNew, yNew, zNew, uxNew, uyNew, uzNew, txWall, tyWall, tzWall, dEm = damCom1DFA.getWallInteraction(x, y, z,
```
Comment (fso42) [inline avaframe/com1DFA/DFAfunctionsCython.pyx:704]: plese add a pytest for this part (at least checking if it runs...)

File: avaframe/com1DFA/com1DFA.py [hunk 3]
```diff
@@ -527,8 +529,22 @@ def prepareInputData(inputSimFiles, cfg):
         resLine = None
         resistanceArea = ''
 
+    # get line from dam
+    if cfg['GENERAL'].getboolean('dam'):
+        if entResInfo['dam'] == 'Yes':
+            damFile = inputSimFiles['damFile']
```
Comment (fso42) [inline avaframe/com1DFA/com1DFA.py:535]: Please a a pytest to see if it runs

File: avaframe/com1DFA/particleTools.py [hunk 4]
```diff
@@ -681,11 +681,11 @@ def findParticles2Track(particles, center, radius):
     track = True
     x = particles['x']
     y = particles['y']
-    z = particles['z']
+    # z = particles['z']
```
Comment (fso42) [inline avaframe/com1DFA/particleTools.py:684]: just remove these?

File: avaframe/in3Utils/geoTrans.py [hunk 5]
```diff
@@ -701,6 +706,80 @@ def isCounterClockWise(path):
     return isCounterClock
 
 
+def getCellsAlongLine(header, lineDict, addBuffer=True):
+    """ Find all raster cells crossed by the line
+    line has to be entierly contained on the raster extend. If addBuffer is True, add neighbour cells to the result
+    based on https://stackoverflow.com/a/35808540/15887086
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:712]: Please expand the docstring

**Review by fso42 — CHANGES_REQUESTED** [review]
General remark: any deeper reason why you use an indent of 2 (instead of our normal 4) spaces for the cython files?

File: avaframe/in3Utils/geoTrans.py [hunk 6]
```diff
@@ -701,6 +706,80 @@ def isCounterClockWise(path):
     return isCounterClock
 
 
+def getCellsAlongLine(header, lineDict, addBuffer=True):
```
Comment (awirb) [inline avaframe/in3Utils/geoTrans.py:709]: general question, is it an option to make a buffered line and look for the affected cells, maybe shapely offers tools for this, similar to the create ReleaseBuffer function..

File: docs/damToolCom1DFA.rst [hunk 7]
```diff
@@ -0,0 +1,92 @@
+Dam tool com1DFA
+=================
+
+The com1DFA computational module provides the option to take the effect of dams into account.
+This is done using a haddock method based on particles bouncing on a dam wall.
```
Comment (awirb) [inline docs/damToolCom1DFA.rst:5]:
```suggestion
This is done using an ad-hoc method based on particles bouncing on a dam wall.
```

Comment (awirb) [reply docs/damToolCom1DFA.rst:5]: maybe bouncing back or being reflected/deflected by a dam wall?

Comment (awirb) [reply docs/damToolCom1DFA.rst:5]: and maybe based on the idea of particles ---

File: docs/damToolCom1DFA.rst [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+Input
+-------
+
+The dam is described by the crown line, the slope and the restitution coefficient:
+
+
+  * crown line as shape file (use the line type and enable the enable the "additional dimentions" option in order
```
Comment (awirb) [inline docs/damToolCom1DFA.rst:13]:
```suggestion
  * crown line as shape file (use the line type and enable the "additional dimensions" option in order
```

File: docs/damToolCom1DFA.rst [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    to specify the z coordinate). The z coordinate corresponds to the absolute height. The dam is then located on the
```
Comment (awirb) [inline docs/damToolCom1DFA.rst:14]:
```suggestion
    to specify the z coordinate). The z coordinate corresponds to the absolute height (terrain elevation plus dam height). The dam is then located on the
```

File: docs/damToolCom1DFA.rst [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    left side of the dam. The dam shape files lives in the ``avaDir/Inputs/DAM/`` directory.
```
Comment (awirb) [inline docs/damToolCom1DFA.rst:15]:
```suggestion
    left side of the dam. The dam shape files live in the ``avaDir/Inputs/DAM/`` directory. 
```

Comment (awirb) [reply docs/damToolCom1DFA.rst:15]: only one allowed? maybe add this info

File: docs/damToolCom1DFA.rst [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+(slope measured from the horizontal, :math:`\beta`) and a restitution coefficient (describing if we consider
+more elastic or inelastic collisions between the particles and the dam wall, varying between 0 and 1).
+
+The geometrical description of the dam is given on the figure :numref:`fig-DamToolSide`.
+The dam crown line (:math:`\mathbf{x_\text{crown}}`) is projected onto the topography, which provides us
+with the dam center line (:math:`\mathbf{x_\text{center}}`). We compute the tangent vector to the
+center line (:math:`\mathbf{t_f}`). From this tangent vector and the dam slope, it is possible to
+compute the wall tangent vector (:math:`\mathbf{t_w}`). Knowing the wall tangent vector and height,
+it is possible to determine normal vector to the wall (:math:`\mathbf{n_w}`) and the foot line which
+is intersection between the dam wall and the topography (:math:`\mathbf{x_\text{foot}}`).
```
Comment (awirb) [inline docs/damToolCom1DFA.rst:40]:
```suggestion
is the intersection between the dam wall and the topography (:math:`\mathbf{x_\text{foot}}`).
```

File: docs/damToolCom1DFA.rst [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+When the dam fills up (flow thickness increases), the foot line is modified
+(:math:`\mathbf{x_\text{foot}^\text{filled}} = \mathbf{x_\text{foot}} + \frac{h_v}{2} \mathbf{e_z}`).
+The normal and tangent vectors to the dam wall is readjusted accordingly.
```
Comment (awirb) [inline docs/damToolCom1DFA.rst:44]:
```suggestion
The normal and tangent vectors to the dam wall are readjusted accordingly.
```

File: docs/damToolCom1DFA.rst [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+
+.. _fig-DamToolSide:
+
+.. figure:: _static/damToolSideView.png
+          :width: 90%
+
+          Side view of the dam (cut view). :math:`\mathbf{x_\text{crown}}` describe the crown
```
Comment (awirb) [inline docs/damToolCom1DFA.rst:52]:
```suggestion
          Side view of the dam (cut view). :math:`\mathbf{x_\text{crown}}` describes the crown
```

File: docs/damToolCom1DFA.rst [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+          of the dam, :math:`\mathbf{x_\text{center}}` is the vertical projection of the crown
+          on the topography (here the light blue line represents the topography).
+          The tangent vector to the center line (:math:`\mathbf{t_f}`) is computed from the
+          center line points. The tangent vector to the center line with the dam slope angle enable to compute
+          the tangent (:math:`\mathbf{t_w}`) and normal (:math:`\mathbf{n_w}`) vector to the dam wall.
+          Finally, this normal vector is adjusted depending on the snow thickness at the dam location
+          (filling of the dam)
```
Comment (awirb) [inline docs/damToolCom1DFA.rst:59]: (filling of the dam, n_w _filled).

**Review by awirb — CHANGES_REQUESTED** [review]

File: avaframe/out3Plot/outDebugPlots.py [hunk 15]
```diff
@@ -2,7 +2,7 @@
 import pathlib
 import copy
 import matplotlib
-matplotlib.use('agg')
+# matplotlib.use('agg')
```
[the remark this answers is not in the document]
Comment (fso42) [reply avaframe/out3Plot/outDebugPlots.py:5]: Ah no, please either remove or leave it commented. See issues #745 and #573

File: avaframe/in3Utils/geoTrans.py [hunk 16]
```diff
@@ -30,6 +31,10 @@ def projectOnRaster(dem, Points, interp='bilinear'):
         Points dictionary (x,y)
     interp: str
         interpolation option, between nearest or bilinear
+    what: str
+        which key in the dem dict should be used for the interpolation?
+    where: str
+        which key in the Points dict should be updated with the interpolated data?
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:37]: mh, not sure whether what/where are the best names. from/to ? source/target? sourceKey/targetKey?

**Review by fso42 — CHANGES_REQUESTED** [review]
General remark: the pyx files have a ident setting of 2, whereas the rest of our code has 4. Please check whether you can easily reindent with your editor. Otherwise let me know and I'll try... (But I'd prefer you showing up as the contributor for those lines :-) )

**Pull request merged**